File: day03.py
with open('day03.input') as f:
    s = f.read()

d = {'>': 1, '<': -1, '^': 1j, 'v': -1j}

for part in 1, 2:
    p, loc = {0}, [0, 0]
    for i, m in enumerate(s):
        loc[i % part] += d[m]
        p.add(loc[i % part])
    print(len(p))

File: 2015/day03/day03.nim
import basic2d, sets

let s = readFile "day03.input"

for j in 1..2:
  var loc = [new(ref Point2d), new(ref Point2d)]
  var p = initSet[string]()
  p.incl("0,0")
  for i, m in s:
    var sr = loc[i mod j]
    case m:
      of '>':
        sr[].move(1, 0)
      of '<':
        sr[].move(-1, 0)
      of '^':
        sr[].move(0, 1)
      of 'v':
        sr[].move(0, -1)
      else:
        discard
    p.incl($sr[])
  echo len p
